Fix slither issues

yerdua opened the pull request:
### Description

This fixes many of the issues that came out of a Slither analysis.

### Tested

I reran the slither analysis and got this result:
[slither-report-2019-09-06.txt](https://github.com/celo-org/celo-monorepo/files/3584430/slither-report-2019-09-06.txt)

### Related issues

- Fixes #372

yerdua committed:
based on the slither report, fixed a bunch of uses of `public` that
should have been `external`, and some things that should have been
constants.

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
use reentrancy guard

m-chrzan commented on packages/protocol/contracts/identity/Attestations.sol:
```diff
@@ -149,7 +150,7 @@ contract Attestations is IAttestations, Ownable, Initializable, UsingRegistry {
    * be paid.
    */
-  function request(
+  function nonReentrant request(
```
Modifier should go after `external`. I would have expected this to be picked up by the compiler and/or linter?

yerdua replied:
I forgot to run the linter. This is fixed in a newer commit.

m-chrzan reviewed (commented):
Looks good so far!

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
change pragma version. fix some other things that slither reported on

yerdua commented on packages/protocol/contracts/Migrations.sol:
```diff
@@ -1,4 +1,4 @@
-pragma solidity ^0.5.8;
+pragma solidity ^0.5.3;
```
This was one of the changes suggested in the slither report, selected by @asaj as an issue that should be fixed. However, version-changes are rarely free of consequences. Is this a good idea?

m-chrzan replied:
Did the slither report say anything about why this is the preferred version?

Behavior wise, I think in our case it should be fine. I don't think we're using any features related to minor version changes. If we are, regression tests should hopefully catch that ;)

yerdua replied:
This was the message:
`Pragma version "^0.5.8" necessitates versions too recent to be trusted. Consider deploying with 0.5.3`

yerdua commented:
I updated the issue #372 with the things I wasn't able to figure out.

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
switched some functions from `public` to `external`, and switched `memory` to `calldata`

yerdua committed:
use ReentrancyGuard properly

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
update expected gas usage in e2e transfer tests

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
switch another function from public to external

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua commented:
New slither report is here: [slither-report-2019-09-06.txt](https://github.com/celo-org/celo-monorepo/files/3584430/slither-report-2019-09-06.txt)

While most of the issues that @asaj highlighted for fixing are no longer appearing here, there's a few confusing ones remaining. The initial report contained issues for `BondedDeposits` which has since been renamed to `LockedGold`. This new report detects Reentrancy in some functions in `governance/LockedGold.sol`. However, _all_ of these have the `nonReentrant` modifier that fixed this problem in other contracts. Is there still something wrong, or is it possible that slither is reporting false-positives? (cc: @m-chrzan)

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

m-chrzan commented on packages/protocol/contracts/stability/Exchange.sol:
```diff
@@ -293,7 +295,7 @@ contract Exchange is IExchange, Initializable, Ownable, UsingRegistry {
    * the price reported by the Oracle.
    */
-  function _updateBucketsIfNecessary() private {
+  function _updateBucketsIfNecessary() private nonReentrant {
```
This worries me. `exchange` can call this function. Calling one `nonReentrant` function from another is not supported (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/ReentrancyGuard.sol#L27). 

We should add unit tests that test calling `exchange` in a situation where buckets should be updated.

yerdua replied:
so... _should_ `exchange` be calling this function?

I'm not clear on what exactly this problem is and how to handle it. My rough sense of what you're saying here is that adding the `nonReentrant` modifier is the wrong thing. But, since the reentrancy bug is still a thing we care about preventing, we'd have to do something else?

m-chrzan replied:
Yes. Exchange should still be calling this.

>But, since the reentrancy bug is still a thing we care about preventing, we'd have to do something else?

In most of these cases, though Slither reports a *potential* reentrancy problem, we can be fairly confident that nothing will happen. Reentrancy is a problem when the called contract is untrusted. In most (if not all) of our cases, we're calling other contracts in the Celo system.

m-chrzan commented on packages/protocol/contracts/stability/Exchange.sol:
```diff
@@ -236,7 +238,7 @@ contract Exchange is IExchange, Initializable, Ownable, UsingRegistry {
     * @param newMininumReports The new update minimum number of reports required
     */
-  function setMinimumReports(uint256 newMininumReports) public onlyOwner {
+  function setMinimumReports(uint256 newMininumReports) external onlyOwner {
```
`setUpdateFrequency` above is still `public`.

Although I'm thinking that instead we should make these `public` and use them in `initialize` (and add setters for the parameters that don't have them yet).

yerdua replied:
In my latest commit, I've switched this back to `public` and updated initialize() to use the setters instead of directly setting the variables.

The remaining `initialize` params without setters are these:
```
address stableToken,
uint256 spreadNumerator,
uint256 spreadDenominator,
uint256 reserveFractionNumerator,
uint256 reserveFractionDenominator,
```

`stableToken` is used directly to set the value for `stable`. The other ones make up fractions.

Can you tell me more about why to prefer setters over setting the values with an `=` in the initializer? Are there any cases we wouldn't want this? Am I right to assume that adding more of these setter would also necessitate adding and emitting events?

Also related, what's up with some of the params starting with `_` and others not? Currently the ones starting with `_` have setters and the ones without do not. If I make setters for the rest of them, do they get underscores as well?

m-chrzan replied:
I think most of why there's some inconsistencies here is because I rushed a bit when originally getting this code in, sorry! 

We want setters so that the parameters are easily controllable via governance. In the setters we should emit events, and that's why we don't want to set these values directly in `initialize` - to deduplicate all the additional logic that happens when setting a parameter (e.g. emitting an event).

I use `_` for parameters that would otherwise have the same name as a storage variable. Since fraction elements get put into a struct, but have to be provided separately as arguments, there's no clash. Although this is going to change when #337 makes it in.

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
remove nonReentrant modifier from updateBucketsIfNecessary because
it needs to be called by another function with that modifier.

Use setters in Exhange initialization

yerdua committed:
set stable token address in a setter and emit an event

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
more setters, fix gas fees

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua commented on packages/protocol/contracts/stability/Exchange.sol:
```diff
@@ -37,7 +38,20 @@ contract Exchange is IExchange, Initializable, Ownable, UsingRegistry {
+  );
+
+  event SpreadSet(
+    uint256 spread
```
@m-chrzan `spread` and `reserveFraction` are of type `Fixidity.Fraction`, but this doesn't seem to be allowed in events. It resulted in a TypeError during compilation when I tried to use it here:

```
TypeError: This type is only supported in the new experimental ABI encoder. Use "pragma experimental ABIEncoderV2;" to enable the feature.
    FixidityLib.Fraction spread
```

Since the param given to `initialize` is a `uint256`, I went with that. Is that the right thing to do?

m-chrzan replied:
Yep, that's exactly the way to do it! You can use `.unwrap()` on a Fixidity Fraction to get the internal representation.

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

m-chrzan commented on packages/protocol/test/stability/exchange.ts:
```diff
@@ -334,6 +392,22 @@ contract('Exchange', (accounts: string[]) => {
+          }
+        )
+        const newStableBalance = await stableToken.balanceOf(user)
+        assertEqualBN(newStableBalance, expectedStableBalance)
```
Can you verify the buckets updated after this? Wouldn't the expected balance be different after a bucket update?

m-chrzan commented on packages/protocol/test/stability/exchange.ts:
```diff
@@ -211,6 +211,64 @@ contract('Exchange', (accounts: string[]) => {
     })
   })
 
+  describe('#setSpread', () => {
```
Sweet, thanks for the tests! Looks like `setStableToken` still needs one?

yerdua replied:
whoops. Thanks!

yerdua committed:
stable token test (start)

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
finish/fix tests for setStableToken

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
tries to write a test that verifies buckets were updated and got stuck on figuring out what's supposed to happen

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
make the exchange tests have the right values

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

m-chrzan commented on packages/protocol/test/stability/exchange.ts:
```diff
@@ -211,6 +211,94 @@ contract('Exchange', (accounts: string[]) => {
+      await exchange.setStableToken(newStable)
+
+      const actualStable = await exchange.stable()
+      console.info('actual stable:', actualStable)
```
Remove logs from test.

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
remove console.log statements

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues

yerdua committed:
Merge branch 'master' into yerdua/fix-slither-issues
